### PR TITLE
add tcp transport

### LINKED ISSUE #107: Upgrade ZIO to 1.0.0-RC17
Fix all errors present in the current codebase that prevent automatic migration.

### PR DESCRIPTION
**Pull request by mschuwalow**
closes #107

Initial version of tcp based transport

### CODE REVIEW

File: membership/src/main/scala/zio/membership/Error.scala [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+import zio.duration.Duration
+import zio.nio.SocketAddress
 
 // TODO: define error hierarchy
-sealed trait Error extends Exception
+sealed abstract class Error(msg: String = "", cause: Throwable = null) extends Exception(msg, cause)
 
-sealed trait SendError    extends Error
-sealed trait ReceiveError extends Error
+sealed abstract class SendError(msg: String = "", cause: Throwable = null) extends Error(msg, cause)
```
Comment (pshemass) [inline membership/src/main/scala/zio/membership/Error.scala]: We could simplify this to ```TransportError```

```
sealed abstract class TransportError(msg: String = "", cause: Throwable = null) extends Error(msg, cause)

final case class ExceptionThrown(exc: Throwable) extends TransportError(cause = exc)
final case class RequestTimeout(timeout: Duration) extends TransportError(msg = s"Request timeout $timeout.")
final case class ConnectionTimeout(timeout: Duration) extends TransportError(msg = s"Connection timeout $timeout.")
final case class BindFailed(addr: SocketAddress, exc: Throwable)
       extends TransportError(msg = s"Failed binding to address $addr.", cause = exc)
 
```

File: membership/src/main/scala/zio/membership/transport/tcp.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                                  } yield data
+                                )
+                                .timeout(connectionTimeout)
+                                .catchAllCause(_ => ZStream.empty)
+                            }
+                          }
+                        }
+                        .mapError(ReceiveError.ExceptionThrown)
+                      // hack to properly handle interrupts
+                      messages.merge(ZStream.never.ensuring(close))
```
Comment (pshemass) [inline membership/src/main/scala/zio/membership/transport/tcp.scala]: can we add test for interrupts?

File: build.sbt [hunk 3]
```diff
@@ -58,12 +64,13 @@ lazy val membership = project
...
-      "dev.zio"     %% "zio-streams"  % "1.0.0-RC16",
-      "dev.zio"     %% "zio-nio"      % "0.3.0",
-      "com.lihaoyi" %% "upickle"      % "0.8.0",
-      "dev.zio"     %% "zio-test"     % "1.0.0-RC16" % "test",
-      "dev.zio"     %% "zio-test-sbt" % "1.0.0-RC16" % "test"
+      "dev.zio"     %% "zio"             % "1.0.0-RC17",
+      "dev.zio"     %% "zio-streams"     % "1.0.0-RC17",
+      "dev.zio"     %% "zio-nio"         % "0.4.0",
+      "dev.zio"     %% "zio-macros-core" % "0.6.0",
+      "com.lihaoyi" %% "upickle"         % "0.8.0",
```
[the remark this answers is not in the document]
Comment (mijicd) [reply build.sbt]: Chosen as the most stable among the existing ones. There were a couple of other candidates, but we put them aside for various reasons (at least until we come up with our own codec library :)).

File: membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+            environment >>> Live.live(for {
+              port   <- freePort
+              addr   <- SocketAddress.inetSocketAddress(port)
+              chunk  <- readOne(addr).fork
+              _      <- send(addr, payload).delay(connectDelay)
+              result <- chunk.join
+            } yield assert(result, equalTo(payload)))
+          }
+        } @@ flaky @@ timeout(120.seconds) // this one is slow
```
Comment (mijicd) [inline membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala]: We should probably follow ZIO's approach and allocate machine with more resources for tests like this one. Also, we can consider having an alternate test strategy for long-running tests (i.e. not on every PR).

However, that should be done in the follow-up.

File: membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import zio.clock.Clock
+import zio.duration._
+import zio.macros.delegate._
+import zio.nio.SocketAddress
+import zio.test.environment.TestClock
+import zio.test.environment.Live
+import zio.test.TestAspect._
+
+object TransportSpec
+    extends DefaultRunnableSpec({
```
Comment (mijicd) [inline membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala]: I'm looking forward to having inheritance-based specs here.

File: membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      // todo: actually find free port
+      def freePort = ZIO.succeed(8010)
```
Comment (mijicd) [inline membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala]: It can be val.

**Comment by mijicd** [comment]
closes #107

File: membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala (hunk 5 above)
Comment (mschuwalow) [reply membership/src/test/scala/zio/membership/transport/TcpTransportSpec.scala]: Yeah, they are much less awkward

File: build.sbt (hunk 3 above)
[the remark this answers is not in the document]
Comment (pshemass) [reply build.sbt]: that's great! writing those serializers are unnecessary burden at this stage so I agree that we should use something for that.

**Pull request merged**